api: drivers: serial: uart_async: Call UART_RX_RDY event after rx_disable()

nordic-krch opened the pull request:
As proposed in #25317 modified API description and updated the test and nrf implementations.

Short summary:
Before: calling `rx_disable()` stops receiver and discards any pending, received data.
After: after calling`rx_disable()` `UART_RX_RDY` event is generated if there is any pending data.

PR contains changes in all 3 drivers which implemented asynchronous API.

nordic-krch committed:
drivers: uart: nrf_uarte: Generate RX_RDY after rx_disable

So far, when rx_disable was called then received data was discarded.
This is currently not according to the API but it is needed if
rx_disabled is called due to out of band information about end of
packet.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: uart: RX_RDY after rx_disable

Updated rx_disable() description so that RX_RDY event is generated
after calling rx_disable. It contains data received prior to abort.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: Update uart_async read_abort test

Updated test to reflect API change.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: uart: nrf_uart: Generate RX_RDY after rx_disable

So far, when rx_disable was called then received data was discarded.
This is currently not according to the API but it is needed if
rx_disabled is called due to out of band information about end of
packet.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
There is one more implementation of this API: `uart_sam0.c` added by @Sizurka. Any chance to update that implementation? As you can see these are usually small tweaks.

Sizurka commented:
I believe that's how I already made the SAM0 implementation work: [uart_sam0_notify_rx_processed](https://github.com/zephyrproject-rtos/zephyr/blob/c61dfdc680615a1d636d1a676897815860ca4da5/drivers/serial/uart_sam0.c#L993).  Except that I released the inactive (if any) buffer before the ready event (i.e. `UART_RX_BUF_RELEASED` (inactive), `UART_RX_RDY` (active), `UART_RX_BUF_RELEASED` (active), `UART_RX_DISABLED`).  So if the new test case enforces that, then the release would need to be re-ordered.

That said, that particular path was untested and I no longer have a functional Zephyr environment, so I can't verify it against the new test case.

nordic-krch committed:
drivers: uart: sam0: Reorder events on rx_disable

Modified order of reported events on rx disable to match API
description: first RX_RDY and then RX_BUF_RELEASED.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@Sizurka can you take a look at https://github.com/zephyrproject-rtos/zephyr/pull/25967/commits/3d71e790b9acb6d2f3ea0b5694211f232444221a where I've modified the order. I will try to find someone with this board to run the test.

anangl commented on include/drivers/uart.h:
```diff
@@ -551,8 +551,9 @@ static inline int uart_rx_buf_rsp(struct device *dev, u8_t *buf, size_t len)
- * UART_RX_BUF_RELEASED event will be generated for every buffer scheduled,
- * after that UART_RX_DISABLED event will be generated.
+ * UART_RX_RDY will be generated if there is pending received data followed by
+ * UART_RX_BUF_RELEASED event for every buffer scheduled, after that
```
This should be rephrased. Currently this "pending received data followed by UART_RX_BUF_RELEASED event" is quite confusing.
In addition to the mention here, the description of the [UART_RX_RDY](https://github.com/zephyrproject-rtos/zephyr/blob/3d71e790b9acb6d2f3ea0b5694211f232444221a/include/drivers/uart.h#L99) event should be updated (currently it mentions only two cases in which this event is generated). Also the "Receiving" section, in part that [describes](https://github.com/zephyrproject-rtos/zephyr/blob/3d71e790b9acb6d2f3ea0b5694211f232444221a/include/drivers/uart.h#L67) the UART_RX_STOPPED and UART_RX_BUF_RELEASED events, could mention that before a given buffer is released, the UART_RX_RDY event is generated for any data received to that buffer that has not yet been passed to the user callback.

nordic-krch replied:
can you take another look?

anangl replied:
I still find it more confusing than explaining. What about just adding a sentence about the UART_RX_RDY event after the description that we've got now? I mean something like:
```
 * Additionally, if there is any pending received data, the UART_RX_RDY event
 * for that data will be generated before the UART_RX_BUF_RELEASED events.
```

anangl reviewed (commented):
I think that there is no need to change the order of events generated by the uart_sam0 driver. As long as `UART_RX_RDY` is generated before `UART_RX_BUF_RELEASED` for the same buffer, it should be fine. It should not matter if the second buffer is released before or after that.

nordic-krch commented:
> I think that there is no need to change the order of events generated by the uart_sam0 driver. As long as UART_RX_RDY is generated before UART_RX_BUF_RELEASED for the same buffer, it should be fine. It should not matter if the second buffer is released before or after that.

nordic-krch commented:
> I think that there is no need to change the order of events generated by the uart_sam0 driver. As long as UART_RX_RDY is generated before UART_RX_BUF_RELEASED for the same buffer, it should be fine. It should not matter if the second buffer is released before or after that.

I think that it would be good to specify that and have it consistent. We have only 3 implementation so it is still quite early. I was thinking that user may expect that buffer will be released in the same order as the were provided.

carlescufi commented:
@nzmichaelh and @pabigot could you test the changes on sam0?

nordic-krch committed:
doc: Add CONFIG_UART_ASYNC_API to the doxygen defines

Added CONFIG_UART_ASYNC_API to enable documentation generation for
asynchronous API.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

carlescufi commented:
@anangl can you revisit?
